### ISSUE TITLE
Ability to disable shadow depth texture computation for ProjectedTexture (FPS saver)

### ISSUE THREAD

**Issue by maurits150** [Facepunch/garrysmod-requests#2137]
I am trying to render a mesh based world that looks like this:

<img src="https://maurits.tv/data/img/July%202023/2023-07-20_22-25-35_whkiB4Khsd.png" width="600">

Rendering the mesh is a bit more expensive compared to a brush based world but I am able to make it work good enough.

In order to light this world I want to use a ProjectedTexture (practically the only way to do this without it looking like garbage).

A ProjectedTexture with shadows enabled stores its depth information in a texture called _rt_shadowdepthtexture_<0-7>. 
This looks like this for my world:

<img src="https://maurits.tv/data/img/July%202023/2023-07-20_22-29-21_rRSbgtvfh4.png" width="600">

This texture seems to be rendered using the standard CViewSetup (kinda like render.RenderView).  
This happens every frame, causing each dynamic shadow to render the entire world again but from the lamp's perspective. 
This is why dynamic shadows have a huge performance impact.

Engine code: https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/client/clientshadowmgr.cpp#L3949

In my case I am rendering a static world light that does not move and the shadows do not need to be recomputed often. 
Rendering the entire (mesh) world each frame just to generate a static shadowdepthtexture is a waste of CPU time.

What I am requesting is the ability to disable ComputeShadowDepthTextures for a specific ProjectedTexture object. The allocated shadowdepthtexture will simply keep the last rendered depthmap texture in memory. This way I am able to keep casting shadows while not updating them in real-time and saving lots of CPU time.

The function to change this could e.g. be called `SetNextShadowComputation` (similar to some NextThink functions).
It accepts the time at which the next shadow computation should take place. Setting it to CurTime() + 1 will make the flashlight update after 1 second. The default value will be 0 which is always below CurTime() and will cause shadows to be always computed by default.

The timestamp can be stored in the flashlightState. This state that already contains the shadow depth biases which are accessible in lua. The loop simply needs to `continue` a specific light as long as the set value is above CurTime().

**Comment by maurits150** [comment]
Optionally, the ability to override this depth texture entirely for a specific ProjectedTexture would also be great.
I would love to create my own depth texture and then populate it with render.RenderView. In that case I could also bypass the current hardcoded engine limit of 8 textures (which is probably just there to avoid rendering the world more than 8 times? memory usage is not really a problem these days).

**Comment by maurits150** [comment]
Optionally, the ability to throttle this update rate in the entire game using a console command would allow game wide FPS savings at the cost of a bit more choppy shadows (at say 20 FPS it would still be playable?)

**Comment by maurits150** [comment]
It looks like it's also possible to cache the shadow vertices between frames.

https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/game/client/clientshadowmgr.cpp#L1856

If this implies that we can cache the actual vertex cache.. the one render.RenderFlashlights seems to build and render out:

<img src="https://maurits.tv/data/img/July%202023/2023-07-21_20-49-46_nAd0xezjLA.png" width=600>

Then we could make static shadows with close to zero resource usage? Not sure if my assumptions are correct here

**Comment by ZH-Hristov** [comment]
IIRC Portal 2's branch has projected texture caching, but the code for that is probably not public. ASW branch might have it though.

**Comment by maurits150** [comment]
I kinda managed to assault the engine into doing this by setting `SetEnableShadows`+`Update` the lamp to `true` before render.RenderFlashlights and then setting it back to `false` after rendering flashlights. 

Here I put two pallets next to each other then disabled updating the depth texture using this trick, and the shadow becomes frozen.

![Assault](https://maurits.tv/data/img/August%202023/2023-08-13_22-42-45_rmiUbfJxk0.png)

The only issue is that if there's any other other lamps the RT gets overwritten by those again, so still kind of hard to do this reliably.

**Comment by maurits150** [comment]
> IIRC Portal 2's branch has projected texture caching, but the code for that is probably not public. ASW branch might have it though.

Yea I guess it is this from the https://developer.valvesoftware.com/wiki/Env_projectedtexture wiki:

https://maurits.tv/data/img/September%202023/2023-09-15_00-30-18_TuUSlxEPA6.png

Lightstyle was backported so maybe this one can be as well? 
Although I am not sure if I can use the actual entity because I want to render this thing clientside only.

**Comment by ZH-Hristov** [comment]
Yeah it'd need to be ported to GMod's projected textures, which are not "entities" per se

**Comment by Jaffies** [comment]
MUST HAVE

**Comment by maurits150** [comment]
Having a crack at this again. Figured that merging meshes into a single one helps more than expected.. a few thousand draw calls is pretty heavy. So my individual world building blocks are now merged into a single mesh (per material and per max 64k tri's).

Still, this feature would be great to get my FPS back into the 300+ range. It would also allow me to add more 'quality' levels without sacrificing more FPS, by using more than a single projected textures, like in this example.

 ![alt text](https://maurits.tv/data/img/March%202024/2024-03-02_13-31-41_b74MdjABxe.png)

There are workshop mods that do this same trick to achieve dynamic world lighting. Having this feature would also make these mods much more efficient. 

Real CSM - Modern Lighting & Realistic Dynamic Shadows
https://steamcommunity.com/sharedfiles/filedetails/?id=2817072898

Just having the dynamic lights not update the render target (and render the whole world) every frame saves on rendering time. And being able to cache the mesh it projects into the world (like portal 2) would save even more FPS. I already mentioned where it's found in the engine code last year.
